SonarScanner.Shim/PropertiesFileGenerator.cs

```diff
@@ -102,48 +102,125 @@ public bool TryWriteProperties(PropertiesWriter writer, out IEnumerable<ProjectD
+        {
+            var fileWithProjects = projects
+                .SelectMany(p => p.ReferencedFiles.Select(f => new { Project = p, File = f }))
+                .GroupBy(group => group.File, StringComparer.OrdinalIgnoreCase)
+                .ToDictionary(group => group.Key, group => group.Select(x => x.Project).ToList());
+
+            foreach (var group in fileWithProjects)
+            {
+                var file = group.Key;
+
+                if (!File.Exists(file))
                 {
-                    writer.WriteSettingsForProject(projectData,
-                        projectData.CoverageAnalysisExists(logger) ? projectData.VisualStudioCoverageLocation : null);
+                    logger.LogWarning(Resources.WARN_FileDoesNotExist, file);
+                    logger.LogDebug(Resources.DEBUG_FileReferencedByProjects, string.Join("', '",
+                        group.Value.Select(x => x.Project.GetProjectDirectory())));
+                    continue;
                 }
 
-                // Handle global settings
-                writer.WriteGlobalSettings(analysisProperties);
+                if (!PathHelper.IsInFolder(file, rootProjectBaseDir)) // File is outside of the SonarQube root module
+                {
+                    logger.LogWarning(Resources.WARN_FileIsOutsideProjectDirectory, file);
+                    logger.LogDebug(Resources.DEBUG_FileReferencedByProjects, string.Join("', '",
+                        group.Value.Select(x => x.Project.GetProjectDirectory())));
+                    continue;
+                }
 
-                return true;
+                if (group.Value.Count >= 1)
+                {
+                    var closestProject = GetClosestProject(file, group.Value);
+
+                    if (closestProject == null)
+                    {
+                        yield return file;
+                    }
+                    else
+                    {
+                        closestProject.SonarQubeModuleFiles.Add(file);
+                    }
+                }
             }
-            else
+        }
+
+        private void PostProcessProjectStatus(IEnumerable<ProjectData> projects)
+        {
+            foreach (var project in projects)
             {
-                logger.LogError(Resources.ERR_NoValidProjectInfoFiles);
+                if (project.SonarQubeModuleFiles.Count == 0)
+                {
+                    project.Status = ProjectInfoValidity.NoFilesToAnalyze;
+                }
+            }
+        }
 
-                return false;
+        private static ProjectData GetClosestProject(string filePath, IEnumerable<ProjectData> projects)
+        {
+            var longestMatchingPath = (Length: 0, Items: new List<ProjectData>());
+
+            foreach (var project in projects)
+            {
+                var projectPath = project.Project.GetProjectDirectory();
+                if (filePath.StartsWith(projectPath))
```

Comment:
The directory path returned by _GetProjectDirectory()_ doesn't include a trailing backslash, so a directory like
 _c:\aaa\bbb_
would incorrectly match both of the following:
 _c:\aaa\bbbxxxx\myfile.cs_
 _c:\aaa\bbb.cs_